### DISCUSSION TITLE
Debugging with crazyradio

### DISCUSSION THREAD

**Discussion by Jarvis-X** [bitcraze/discussions#438] [Q&A] [answered]
Hi, crazyfliers!

I am familiar with the firmware structure, and have experience modifying and compiling customized versions of it. I used to include debug.h to debug. The modularity of the fw sometimes requires us to jump between files to track a variable which may take a lot of time. Thus, I have been reading the procedures to perform on-chip debugging for the cf firmware on [this page](https://www.bitcraze.io/documentation/repository/crazyflie-firmware/master/development/openocd_gdb_debugging/). At the beginning of the post, a note states 

> NOTE: Debugging requires our [debug adapter](https://www.bitcraze.io/products/debug-adapter-kit/) and an ST Link V2 Debugger or similar.

I am wondering if it is possible to debug without the adapter? For example, is it possible to debug over the crazyradio, or through direct USB connection?

Best,
Jiawei

**Comment by knmcguire** [comment, marked as answer]
Hi! depends how far you need to go into the debugging part. 
1- On chip debugging with openocd and gdb: you need [the debug adapter](https://www.bitcraze.io/products/debug-adapter-kit/) and an off the shelf programmer like stlink v2 or j-link. 
2- Simple printouts can be done with [DEBUG_PRINT](https://github.com/bitcraze/crazyflie-firmware/blob/e82ba19aedcd75a5a95a94618c00cfe6865d8e89/src/utils/interface/debug.h#L51) and read back on the console over air which might be enough for most usecases

Perhaps point 1 can be done through the NRF overair but this has not been implemented at all and I wouldn't even know where to start 😄 and probably the rest here will call me crazy.

**Reply by Jarvis-X** [reply]
Thank you Kimberly!
